test/unit/io_managers_test.py

```diff
@@ -78,3 +79,91 @@ def test_foreign_key_failure(sqlite_io_manager_fixture):
         foreign_key="(artistid)",
         rowids=[1],
     )
+
+
+def test_extra_column_error(sqlite_io_manager_fixture):
+    """Ensure an error is thrown when there is an extra column in the dataframe."""
+    manager = sqlite_io_manager_fixture
+
+    asset_key = "artist"
+    artist = pd.DataFrame(
+        {"artistid": [1], "artistname": ["Co-op Mop"], "artistmanager": [1]}
+    )
+    output_context = build_output_context(asset_key=AssetKey(asset_key))
+    with pytest.raises(OperationalError):
+        manager.handle_output(output_context, artist)
+
+
+def test_missing_column_error(sqlite_io_manager_fixture):
+    """Ensure an error is thrown when a dataframe is missing a column in the schema."""
+    manager = sqlite_io_manager_fixture
+
+    asset_key = "artist"
+    artist = pd.DataFrame(
+        {
+            "artistid": [1],
+        }
+    )
+    output_context = build_output_context(asset_key=AssetKey(asset_key))
+    with pytest.raises(RuntimeError):
+        manager.handle_output(output_context, artist)
+
+
+def test_nullable_column_error(sqlite_io_manager_fixture):
+    """Ensure an error is thrown when a non nullable column is missing data."""
+    manager = sqlite_io_manager_fixture
+
+    asset_key = "artist"
+    artist = pd.DataFrame({"artistid": [1, 2], "artistname": ["Co-op Mop", pd.NA]})
+    output_context = build_output_context(asset_key=AssetKey(asset_key))
+
+    with pytest.raises(IntegrityError):
+        manager.handle_output(output_context, artist)
+
+
+@pytest.mark.xfail(reason="SQLite autoincrement behvior is breaking this test.")
+def test_null_primary_key_column_error(sqlite_io_manager_fixture):
+    """Ensure an error is thrown when a primary key contains a nullable value."""
+    manager = sqlite_io_manager_fixture
+
+    asset_key = "artist"
+    artist = pd.DataFrame(
+        {"artistid": [1, pd.NA], "artistname": ["Co-op Mop", "Cxtxlyst"]}
+    )
+    output_context = build_output_context(asset_key=AssetKey(asset_key))
+    with pytest.raises(IntegrityError):
+        manager.handle_output(output_context, artist)
```

Comment:
## Background
This test is not failing because when there is a missing value in an integer primary key column, SQLite[ uses an autoincrement function ](https://www.sqlite.org/autoinc.html)to fill in the missing value. In the database the primary key column for this example ends up like: `[1, 2]`:

> On an [INSERT](https://www.sqlite.org/lang_insert.html), if the ROWID or INTEGER PRIMARY KEY column is not explicitly given a value, then it will be filled automatically with an unused integer, usually one more than the largest ROWID currently in use. This is true regardless of whether or not the AUTOINCREMENT keyword is used.

To avoid this behavior, SQLite recommends using [`WITHOUT ROWID` phrase](https://www.sqlite.org/withoutrowid.html):

> The special behaviors associated "[INTEGER PRIMARY KEY](https://www.sqlite.org/lang_createtable.html#rowid)" do not apply on WITHOUT ROWID tables. In an ordinary table, "INTEGER PRIMARY KEY" means that the column is an alias for the rowid. But since there is no rowid in a WITHOUT ROWID table, that special meaning no longer applies. An "INTEGER PRIMARY KEY" column in a WITHOUT ROWID table works like an "INT PRIMARY KEY" column in an ordinary table: It is a PRIMARY KEY that has integer [affinity](https://www.sqlite.org/datatype3.html#affinity).

[SQLite has a parameter](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlite-specific-table-options) for controlling this behavior... great!

Not great. As the quote above says, an Integer Primary Key  in a columns using the WITHOUT ROWID becomes a INT PRIMARY KEY that has `integer affinity`.[ From what I can tell sqlite won't throw an error if a string that can't be cast to int is loaded to the db.](https://www.sqlite.org/datatype3.html#affinity) I was able to load `"abc"` value into an "INTEGER PRIMARY KEY" column using WITHOUT ROWID.

SQLite offers a [STRICT phrase](https://www.sqlite.org/stricttables.html) that might restrict the type affinity rules but[ sqlalchemy doesn't support it yet.](https://github.com/sqlalchemy/sqlalchemy/issues/7398)

**I'm worried some of our tables with noncomposite integer primary keys might have autoincrementing values in them. I don't think this is a dagster branch specific issue as I couldn't find anything in `pudl.load` or `pudl. metadata.classes` that protects against this quirk.**

## Solutions
- We don't want autoincrementing primary keys in our data but we also don't want string values to slip in to integer pk columns.
- I think the easiest solution would be check for null values in primary key integer columns on the pandas side.
- To rely on SQLite we could either bypass sqlalchemy to create tables using the STRICT phrase. I still don't fully understand SQLite's behavior when using STRICT and not using sqlalchemy for creating tables feels janky.
- We could also pick up the sqlalchemy STRICT issue. 

@zaneselvans @zschira @jdangerx what do y'all think? Were y'all aware of this quirk and I'm just a dingus?

Reply:
Good to hear we're mostly handling missing values in the pk columns during the cleaning process. Agree we should create a more systematic solution. Seems like this is out of the scope of the PR. I can create a separate issue with this information. I'll leave this xfail test in here.

I'm not sure what the best solution is here. An option is to use the [`dagster-pandera`](https://docs.dagster.io/integrations/pandera) integration which would allow us to check various constraints. Pandera could check primary keys using unique and null constraints but doesn't handle anything to do with referential integrity. Also the dagster-pandera integration currently only handles pandas dataframes.

I agree we'll need to find a tool that can perform integrity tests on any number of databases if we expand to use other dbs.

Reply:
Created an issue: #2415